pkg/controller/backup_deletion_controller.go

```diff
@@ -442,6 +458,85 @@ func (c *backupDeletionController) deleteResticSnapshots(backup *v1.Backup) []er
 	return errs
 }
 
+func getCSIVolumeSnapshotsInBackup(b *velerov1api.Backup, csiClient snapshotter.SnapshotV1beta1Interface, log *logrus.Entry) ([]snapshotv1beta1api.VolumeSnapshot, []error) {
+	csiVolSnaps := []snapshotv1beta1api.VolumeSnapshot{}
+	errs := []error{}
+
+	selector := metav1.LabelSelector{
+		MatchLabels: map[string]string{
+			velerov1api.BackupNameLabel: label.GetValidName(b.Name),
+		},
+	}
+	listOpts := metav1.ListOptions{
+		LabelSelector: metav1.FormatLabelSelector(&selector),
+	}
+
+	for _, ns := range backup.GetNamespacesInBackup(b) {
+		log.Debugf("Listing CSI volumesnapshots in namespace %s", ns)
+		nsvs, err := csiClient.VolumeSnapshots(ns).List(listOpts)
+		if err != nil {
+			errs = append(errs, err)
+		}
+		if nsvs.Items != nil && len(nsvs.Items) != 0 {
+			log.Debugf("Found %d CSI volumesnapshots in namespace %s", len(nsvs.Items), ns)
+			csiVolSnaps = append(csiVolSnaps, nsvs.Items...)
+		}
+	}
+
+	log.Infof("Found %d CSI volumesnapshots in backup %s", len(csiVolSnaps), b.Name)
+	return csiVolSnaps, errs
+}
+
+func setVSCDeletionPolicyToDelete(vscName string, csiClient snapshotter.SnapshotV1beta1Interface) (*snapshotv1beta1api.VolumeSnapshotContent, error) {
+	snapCont, err := csiClient.VolumeSnapshotContents().Get(vscName, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+	old, err := json.Marshal(snapCont)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to marshall existing volumesnapshotcontent %s json", vscName)
+	}
+
+	snapCont.Spec.DeletionPolicy = snapshotv1beta1api.VolumeSnapshotContentDelete
+	new, err := json.Marshal(snapCont)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to marshall updated volumesnapshotcontent %s to json", vscName)
+	}
+
+	pb, err := jsonpatch.CreateMergePatch(old, new)
```

Comment:
In cases like this where the desired patch is known at compile time, I usually just hand-craft it rather than going through the whole diff process at runtime. In this case something like `{"spec":{"deletionPolicy":"Delete"}}` (I may have the field names not quite right) could replace L491-509

Reply:
That should make this much more simpler. Thanks for pointing that out 👍